File: Algorithm/DynamicProgramming/leetcode/72. Edit Distance/GPT/EditDistance.py
# # 1. 問題分析

# ### 競技プログラミング視点

# * **目的**: 2つの文字列の最小編集距離を求める
# * **制約**: `len(word1), len(word2) <= 500` 程度
# * **計算量要件**: `O(m * n)` が実用的
# * **メモリ効率**: `O(min(m, n))` に削減可能
# * **実行速度**: Python の for ループは遅いので、内包表記や組み込み `min` を活用

# ### 業務開発視点

# * 可読性を最優先 → まずは 2次元 DP 実装
# * 型ヒントを入れて静的解析対応
# * エラーハンドリング（引数の型やサイズの検証）を入れる

# ### Python特有の考慮

# * `list` が最適（deque/heap は不要）
# * DP 更新時に **行を使い回す**ことでメモリ削減
# * 内包表記と組み込み関数の活用

# ---

# # 2. アルゴリズム比較

# | アプローチ                | 時間計算量 | 空間計算量       | 実装コスト | 可読性 | ライブラリ活用              | 備考        |
# | -------------------- | ----- | ----------- | ----- | --- | -------------------- | --------- |
# | 方法A: 2次元DP           | O(mn) | O(mn)       | 低     | 高   | list                 | 最も理解しやすい  |
# | 方法B: 1次元DP (ローリング配列) | O(mn) | O(min(m,n)) | 中     | 中   | list                 | 大規模入力に有利  |
# | 方法C: 再帰 + LRUキャッシュ   | O(mn) | O(mn)       | 高     | 中   | functools.lru\_cache | 可読性は高いが遅い |

# ---

# # 3. 実装方針

# * **選択アプローチ**: 方法B (1次元DP)
# * **理由**:

#   * O(mn) 計算量は必須だが、O(min(m, n)) の省メモリ版にすると実用性が高い
#   * Python 実装でも十分高速
#   * 保守性: コメント付きでわかりやすく書ける

# ---

# # 4. コード実装例

# ```python
from typing import List


class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        """
        編集距離 (Levenshtein Distance)

        Args:
            word1 (str): 最初の文字列
            word2 (str): 比較対象の文字列

        Returns:
            int: 最小編集距離

        Raises:
            TypeError: 入力が文字列でない場合
        """
        # 型検証
        if not isinstance(word1, str) or not isinstance(word2, str):
            raise TypeError("Both inputs must be strings")

        m, n = len(word1), len(word2)

        # 片方が空文字の場合
        if m == 0:
            return n
        if n == 0:
            return m

        # 空間最適化: 常に短い方を n にする
        if n > m:
            word1, word2 = word2, word1
            m, n = n, m

        # dp[j] = word1[0:i] と word2[0:j] の編集距離
        prev = list(range(n + 1))

        for i in range(1, m + 1):
            curr = [i] + [0] * n
            for j in range(1, n + 1):
                if word1[i - 1] == word2[j - 1]:
                    curr[j] = prev[j - 1]
                else:
                    curr[j] = 1 + min(
                        prev[j],      # 削除
                        curr[j - 1],  # 挿入
                        prev[j - 1]   # 置換
                    )
            prev = curr

        return prev[n]

    def minDistance_fast(self, word1: str, word2: str) -> int:
        """
        競技プログラミング向け最適化版
        - 入力検証を省略
        - 空間 O(min(m, n))
        """
        if len(word2) > len(word1):
            word1, word2 = word2, word1

        m, n = len(word1), len(word2)
        prev = list(range(n + 1))

        for i in range(1, m + 1):
            curr = [i] + [0] * n
            for j in range(1, n + 1):
                if word1[i - 1] == word2[j - 1]:
                    curr[j] = prev[j - 1]
                else:
                    curr[j] = 1 + min(prev[j], curr[j - 1], prev[j - 1])
            prev = curr

        return prev[n]
# ```

# ---

# # 5. 計算量まとめ

# * **時間計算量**: `O(m * n)`
# * **空間計算量**: `O(min(m, n))`
# * **Python固有のオーバーヘッド**:

#   * インタプリタ実行のため C++ より遅いが、500×500 程度なら問題なし
#   * list の添字アクセスは O(1) で十分高速

# ---

# # 6. Python固有の最適化観点

# * **内包表記**はここでは不可（ループが必須）
# * **list** が最速、`deque` は不要
# * **組み込み min 関数**を活用
# * **ローリング配列でメモリ削減**
# * **型ヒントを付与** → Pylance で静的解析対応
# * **例外処理**で不正入力を防止

# ---

# ✅ まとめ:

# * 業務開発 → `minDistance` を可読性重視で実装
# * 競技プログラミング → `minDistance_fast` を利用（入力検証省略 & 空間最適化）

File: Algorithm/DynamicProgramming/leetcode/72. Edit Distance/GPT/EditDistance.ts
// # 1. 問題の分析

// ### 競技プログラミング視点

// * **実行速度**: DP による O(n·m) アルゴリズムが最適。最大 500×500=25万計算で十分高速。
// * **メモリ効率**: 2 行だけ保持すれば O(min(n,m)) に削減可能。

// ### 業務開発視点

// * **型安全性**: `string` 入力、戻り値は `number`。
// * **保守性**: `prev` / `curr` で分かりやすく命名し、コメントを充実。
// * **エラーハンドリング**: 型ガード・制約長チェック。

// ### TypeScript特有の考慮点

// * **型推論**: `number[]` を明示。
// * **strict mode**: null/undefined チェックを厳密化。
// * **readonly修飾子**: イミュータブルを保証。

// ---

// # 2. アルゴリズムアプローチ比較

// | アプローチ         | 時間計算量  | 空間計算量       | TS実装コスト | 型安全性 | 可読性 | 備考                  |
// | ------------- | ------ | ----------- | ------- | ---- | --- | ------------------- |
// | 方法A: フルDPテーブル | O(n·m) | O(n·m)      | 低       | 高    | 高   | デバッグ向き、可視化しやすい      |
// | 方法B: 2行DP最適化  | O(n·m) | O(min(n,m)) | 中       | 高    | 高   | 最適なメモリ効率            |
// | 方法C: 再帰+メモ化   | O(n·m) | O(n·m)      | 高       | 高    | 中   | 実装が複雑、スタックオーバーフロー懸念 |

// ---

// # 3. 選択したアルゴリズムと理由

// * **選択したアプローチ**: 方法B（2行DP最適化）
// * **理由**:

//   * O(n·m) 時間で十分高速。
//   * メモリは O(min(n,m)) で安定。
//   * 実装コストも妥当で可読性を保てる。
// * **TypeScript特有の最適化ポイント**:

//   * 関数の戻り値型を厳格に `number` に固定。
//   * 入力の型チェックを型システムで担保。
//   * `readonly string` で副作用を防止。

// ---

// # 4. 実装コード（LeetCode形式）

// ```typescript
// LeetCode の形式に合わせて実装
// Node.js v18, TypeScript strict mode 前提
// 外部ライブラリ禁止、CommonJS/ESM いずれでも可（ここでは CommonJS 前提）

/**
 * word1 を word2 に変換するための最小操作回数を計算する
 * @param word1 - 入力文字列1（小文字英字）
 * @param word2 - 入力文字列2（小文字英字）
 * @returns 最小操作回数
 * @throws {TypeError} 入力が文字列でない場合
 * @throws {RangeError} 入力長が 500 を超える場合
 * @complexity Time: O(n*m), Space: O(min(n,m))
 */
function minDistanceGPT(word1: string, word2: string): number {
    if (typeof word1 !== 'string' || typeof word2 !== 'string') {
        throw new TypeError('Inputs must be strings');
    }
    const n: number = word1.length;
    const m: number = word2.length;
    if (n > 500 || m > 500) {
        throw new RangeError('Input length exceeds limit (500)');
    }

    // 常に word1 の方を長くすることでメモリ O(min(n,m)) に削減
    if (n < m) {
        return minDistanceGPT(word2, word1);
    }

    const prev: number[] = new Array(m + 1);
    const curr: number[] = new Array(m + 1);

    // 初期化: word1 が空のとき
    for (let j = 0; j <= m; j++) {
        prev[j] = j;
    }

    for (let i = 1; i <= n; i++) {
        curr[0] = i;
        for (let j = 1; j <= m; j++) {
            if (word1[i - 1] === word2[j - 1]) {
                curr[j] = prev[j - 1];
            } else {
                const del: number = prev[j] + 1; // 削除
                const ins: number = curr[j - 1] + 1; // 挿入
                const rep: number = prev[j - 1] + 1; // 置換
                // Math.min の代わりに3値比較
                let min: number = del < ins ? del : ins;
                if (rep < min) min = rep;
                curr[j] = min;
            }
        }
        // prev ← curr をコピー（再利用）
        for (let j = 0; j <= m; j++) {
            prev[j] = curr[j];
        }
    }

    return prev[m];
}

// LeetCode 提出形式
export { minDistanceGPT };
// ```

// ---

// # 5. TypeScript固有の最適化観点

// ### 型安全性

// * `string` 型のみ受け付けるため、実行時に数値・配列誤入力を防止。
// * 例外条件を明示 (`TypeError`, `RangeError`)。

// ### コンパイル時最適化

// * `const` / `readonly` による副作用抑止。
// * 配列長を固定し、V8 が最適化しやすい形に。

// ### 保守性・可読性

// * 変数名 `prev` / `curr` でロジックが直感的。
// * コメント付きで挿入/削除/置換の操作を明示。

// ---

// ✅ これで **LeetCode回答フォーマット（`function minDistance(...)`）**
// ✅ **TypeScript strict mode / Node.js v18** に準拠した解答になっています。
